LSpec/LSpec.lean

```diff
@@ -0,0 +1,55 @@
+import Lean
+
+abbrev Rel (α : Type _) := α → Bool 
+
+/-- 
+The basic structure used to encode specifications. 
+It represents a list of tests, with each test consisting of 
+a description `descr`, some value `a`, and a test `rel` for `a`.
+-/
+inductive LSpec
+  | done
+  | more (descr : String) (a : α) (rel : Rel α) (next : LSpec)
+  deriving Inhabited
+
+/-- 
+`it` adds a single test to the testing suite.
+This is the same as the projection `LSpec.more`, but extracting it out 
+allows us to set `next` with the default value `done`.
+-/
+def it (descr : String) (a : α) (rel : Rel α) (next : LSpec := LSpec.done) :=
+  LSpec.more descr a rel next
+
+/-- 
+Prints the result of a test.
+-/
+def printMsg : String × Bool → String
+  | (d, true)  => s!"✓ it {d}"
+  | (d, false) => s!"× it {d}"
+
+/--
+Runs a set of `LSpec` tests and appends the results to 
+another list of results (given as input by the caller).
+-/
+def runLSpec (results : List (String × Bool)) : LSpec → List (String × Bool)
```

Comment:
Is it possible to make this an `optparam`? I tried `results : List (String × Bool) = []` but Lean was unhappy with me, and I didn't look into further.

Reply:
We could say it like
```lean
def runLSpec (results : List (String × Bool) := []) : LSpec → List (String × Bool)
```
But since it's the first argument (and the term of `LSpec` is the second), we'd need to specify its value anyways. So the call would end up being `runLSpec (results := []) t`